docs/development/bidirectional-transfer/BidirectionalDataTransfer.md

```diff
@@ -0,0 +1,85 @@
+correlate to these categories as follows:
+
+| Provider Push                                                                               | Consumer Pull                                              |
+|---------------------------------------------------------------------------------------------|------------------------------------------------------------|
+| Consumer offers the forward channel endpoint; provider offers the response channel endpoint | Provider offers the forward and response channel endpoints |
+
+**In each case, the provider always offers the response channel.**
+
+## The Data Plane
+
+The Data Plane establishes data transfer communication channels and endpoints using a *wire protocol*. There are many
+ways to do this, two of which are described below.
+
+**HTTP Endpoints**
+
+The forward and response channels are separate endpoints. The endpoints may be static, where all messages in a
+particular direction are sent to the same endpoint, which then uses a correlation mechanism to process them, for
+example, `https://test.com/forwardChannel` and `https//test.com/responseChannel`. Or, the endpoints may be dynamic,
+where a path part contains a correlation ID, for example, `https://test.com/transferId/forwardChannel`
+and `https://test.com/transferId/responseChannel`.
+
+**Queues and Pub/Sub**
+
+In this scenario, the forward channel is a *queue* or a pub/sub *topic* while the response channel is a *queue*. This is
+a typical architecture used when designing systems with Message-Oriented-Middleware.
+
+### Required Changes to the Data Plane Framework
+
+The required changes to the Data Plane Framework to support bidirectional data transfers are minimal.
+
+#### Response Endpoint `DataAddress`
+
+The `DataAddress` in the `DataFlowResponseMessage` must contain a `https://w3id.org/edc/v0.0.1/ns/responseChannel`
+property of type `DataAddress`. This `DataAddress` follows the same format as the outer `DataAddress` and represents the
+response channel endpoint. For example, it may contain authorization data the consumer uses to access the response
+channel endpoint.
+
+#### The DataPlaneManager
+
+The `DataPlaneManagerImpl` and its collaborators will need to be refactored to generate response
+channel `DataAddresses`:
+
+- `DataPlaneManagerImpl` must be modified to return an EDR in the case of a provider PUSH. This EDR will only contain
+  a `https://w3id.org/edc/v0.0.1/ns/responseChannel` entry. The manager will delegate to `DataPlaneAuthorizationService`
+  to generate the response.
+- `DataPlaneAuthorizationServiceImpl` must be enhanced to support `responseChannel` generation. This should be keyed off
+  of the transfer type. As part of this process, a `DataPlaneAuthorizationServiceImpl.createEndpointDataReference` must
+  generate a `responseChannel` endpoint by delegating to a new
+  method `PublicEndpointGeneratorService.generateCallbackFor(sourceDataAddress).` Access Tokens can be generated
```

Comment:
What exactly is sourceDataAddress in this case? I stuggle with the synchronicity between push and pull scenario here.

Reply:
To understand this requires specialized EDC knowledge. The sourceDataAddress is the reference to the backend asset being transferred. This address is internal to the EDC deployment and not available externally (e.g. to a consumer). The endpoint generation service is responsible for interpreting the address and mapping a publically available endpoint that is associated with retrieving the data.